File: Python/Next Permutation.py
class Solution:
    # @param num, a list of integer
    # @return a list of integer
    def nextPermutation(self, num):
        if len(num) == 0 or len(num) == 1:
            return num
            
        index = len(num) - 2
        while index >= 0 and num[index] >= num[index + 1] :
            index -= 1
        if index == -1:
            return num[::-1]    
        if index >= 0:
            num = num[:index + 1] + list(reversed(num[index + 1:]))
            for i in range(index +1  , len(num)):
                if num[i] > num[index]:
                    temp = num[i]
                    num[i] = num [index]
                    num[index] = temp
                    break
        return num
            
        

File: Java/src/Next Permutation.java
public class Solution {
    public void nextPermutation(int[] num) {
        if (num.length == 1 || num.length == 0) return;
        int index = num.length - 2;
        for(index = num.length-2;index >= 0;index--){
            if(num[index+1] > num[index]){
                break;
            }
        }
        if (index >= 0) {
            for (int i = num.length - 1; i > index; i--){
                if(num[index] < num[i]){
                    int temp = num[index];
                    num[index] = num[i];
                    num[i] = temp;
                    break;
                }
            }
        }
        
        for(int i = index+1, j = num.length - 1; i<j; i++, j--){
            int temp = num[i];
            num[i] = num[j];
            num[j] = temp;
        }
    }
}